### ISSUE TITLE
Array constants are generated on the host during AD

### ISSUE THREAD

**Issue by gnecula** [google/jax#7093]
I noticed that the function `abstract_arrays.zeros_like_shaped_array` has code like `np.broadcast_to(np.array(0, aval.dtype), aval.shape)` (using regular NumPy). These values will be computed on the host during tracing and then may be passed to the device. Is this the desired behavior? 

For example, when we take gradients of constant functions:
```
x = np.ones((10, 10), dtype=np.float32)
jax.make_jaxpr(jax.grad(lambda x: 42.))(x)
```

we get
```
{ lambda a ; b.
  let 
  in (a,) }
```

where a 10x10 array of 0s is built during tracing and passed to the function as the array `a`.
Wouldn't it be better to compute the array of 0s directly on the device?

```
{ lambda  ; a.
   let  b = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(10, 10) ] 0.0
  in (b,) }
```

Incidentally, this affects things like jax2tf where we may end up with large constants embedded in the resulting graph, when instead their computation can be embedded in the graph.

**Comment by mattjj** [comment]
This was intended at one point, pre-lazy-sublanguage and pre-omnistaging #3300, since [there is special logic for handling broadcasted numpy constants efficiently](https://github.com/google/jax/blob/df3cc0d9801cce71a8ce66306c816b1a7813b12b/jax/lib/xla_bridge.py#L539-L571) in XLA lowering. (Hence why it's written with `np.broadcast_to`, which uses stride tricks for the broadcasted representation, unlike `np.zeros`!) But that special handling is only present for XLA lowering, and not jax2tf lowering.

I think we don't need this logic anymore. There might be edge cases where it's more efficient in eager mode, but I can't think of one, and they're probably not that interesting.

I agree we should replace it with `jnp` operations, since that will benefit jax2tf! I can take that on.

**Comment by gnecula** [comment]
Thanks, I can also do it.

**Comment by gnecula** [comment]
@mattjj I have started on this on #7102. The actual change to use lax instead of np is very small, but there are interactions with jax2tf (fixed) and with numpy 3.6.13 (not yet fixed). I can continue to work on this, just FYI, so that you don't start afresh.
